core/dropdowndiv.js

```diff
@@ -744,5 +744,11 @@ DropDownDiv.repositionForWindowResize = function() {
     DropDownDiv.hide();
   }
 };
-
 exports = DropDownDiv;
+
+exports.testOnly_setGetBoundsInfo = function(getBoundsInfoMock) {
+  goog.setTestOnly();
+  getBoundsInfo = getBoundsInfoMock;
```

Comment:
I think this needs to be changed slightly in order to work for other modules that destructure `getPositionMetrics_`. Also, and you may need to check this, I'm not sure this would affect the export if the variable `getPositionMetrics` is modified after the module is loaded.

In the case of destructuring, since the module under test will have already been loaded, this set for `getBoundsInfo` wont affect the local copy of `getBoundsInfo` that the module being tested has destructured.
In order to resolve this, `getPositionMetrics` should call the mock if it exists.

```
let mockGetBoundsInfo = null;

const getBoundsInfo = function(...) {
  if (mockGetBoundsInfo) {
    mockGetBoundsInfo();
  }
  // ...
}

exports.testOnly_setGetBoundsInfo = function(mock) {
  goog.setTestOnly();
  mockGetBoundsInfo = mock;
}
```

Reply:
@moniika: I'm not sure what you mean by "In the case of destructuring, since the module under test will have already been loaded, this set for getBoundsInfo wont affect the local copy of getBoundsInfo that the module being tested has destructured."

I note that `getBoundsInfo` is a local variable in this module, and each module only gets loaded once (no matter how many times it is required), so there is only `getBoundsInfo` and it can have only one value (at a time).

Perhaps the confusion is that when we write `const {func} = goog.require('module');` we sometimes say that "the destructuring gives us a copy of func"—but this is not so: it gives us only _a copy of the reference_ to the  function object in question.  That _can_ cause problems if our tests want to replace  this reference with a reference to a different function, because the importer will continue to use their unmodified copy, which still points at the original implementation.  But destructuring a reference to `getPositionMetrics_` (were it exported) won't affect what function is called when `getPositionMetrics_` invokes `getBoundsInfo`: it will call whatever value was most recently passed to `testOnly_setGetBoundsInfo`, or the original implementation if `testOnly_setGetBoundsInfo` has never been invoked.

Reply:
Thanks for the clarification Chris, that it's a copy of the reference and not a copy of the function. I think I did make that mistake.

I might have also mixed up function names (as I was writing this) and was considering the situation where if `getBoundsInfo` was exported (which is not actually the case) then it could cause issues if `getBoundsInfo` was destructured.